File: backend/internal/models/audit.go
package models

import "time"

type InspectReport struct {
	ReportID       int           `gorm:"primaryKey;autoIncrement;column:ReportID" json:"id"`
	ReportDate     time.Time     `gorm:"column:ReportDate;autoCreateTime" json:"audit_date"`
	Status         string        `gorm:"column:Status;default:draft" json:"status"`
	EmployeeId     string        `gorm:"column:EmployeeId" json:"employee_id"`
	Location       string        `gorm:"column:Location" json:"location"`
	AuditorName    string        `gorm:"column:AuditorName" json:"auditor_name"`
	Summary        string        `gorm:"column:Summary" json:"summary"`
	Recommendation string        `gorm:"column:Recommendation" json:"recommendation"`
	SubmittedAt    *time.Time    `gorm:"column:SubmittedAt" json:"submitted_at"`
	ReviewerName   string        `gorm:"column:ReviewerName" json:"reviewer_name"`
	ReviewNote     string        `gorm:"column:ReviewNote" json:"review_note"`
	ReviewedAt     *time.Time    `gorm:"column:ReviewedAt" json:"reviewed_at"`
	Rows           []Inspect     `gorm:"foreignKey:ReportID" json:"rows,omitempty"`
	Discrepancies  []Discrepancy `gorm:"foreignKey:SessionID;references:ReportID" json:"discrepancies,omitempty"`
}

func (InspectReport) TableName() string { return "InspectReport" }

type Inspect struct {
	InspectID     string    `gorm:"primaryKey;column:InspectID" json:"inspect_id"`
	InspectDate   time.Time `gorm:"column:InspectDate;autoCreateTime" json:"inspect_date"`
	InspectStatus string    `gorm:"column:InspectStatus;default:pending" json:"inspect_status"`
	FoundQty      int       `gorm:"column:FoundQty" json:"found"`
	Condition     string    `gorm:"column:Condition;default:ปกติ" json:"condition"`
	EmployeeId    string    `gorm:"column:EmployeeId" json:"employee_id"`
	ReportID      int       `gorm:"column:ReportID" json:"report_id"`
	AssetID       string    `gorm:"column:AssetID" json:"asset_id"`
	AssetCode     string    `gorm:"column:AssetCode" json:"asset_code"`
	AssetName     string    `gorm:"column:AssetName" json:"asset_name"`
	ExpectedQty   int       `gorm:"column:ExpectedQty" json:"expected"`
	Note          string    `gorm:"column:Note" json:"note"`
}

func (Inspect) TableName() string { return "Inspect" }

type Discrepancy struct {
	ID        uint   `gorm:"primaryKey;autoIncrement" json:"id"`
	SessionID int    `json:"session_id"`
	AssetCode string `json:"asset_code"`
	AssetName string `json:"asset_name"`
	Type      string `json:"type"`
	Expected  string `json:"expected"`
	Actual    string `json:"actual"`
	Cause     string `json:"cause"`
	Action    string `json:"action"`
}

File: frontend/src/services/https/audit.ts
import { apiFetch } from './index'

const token = () => localStorage.getItem('auth_token')

export interface AuditStats {
  total_assets: number
  pending_audit: number
  audited: number
  discrepancies: number
}

// Maps to backend InspectReport model
export interface AuditSession {
  id: number          // ReportID
  location: string    // Location
  audit_date: string  // ReportDate
  auditor_name: string
  employee_id: string
  status: string      // draft | submitted | approved | rejected
  summary: string
  recommendation: string
  submitted_at: string | null
  reviewer_name?: string
  review_note?: string
  reviewed_at?: string | null
  rows?: AuditRow[]
  discrepancies?: Discrepancy[]
}

// Maps to backend Inspect model
export interface AuditRow {
  inspect_id: string
  session_id: number  // ReportID
  asset_id: string    // AssetID
  asset_code: string
  asset_name: string
  expected: number    // ExpectedQty
  found: number       // FoundQty
  condition: string
  note: string
  employee_id: string
}

export interface Discrepancy {
  id: number
  session_id: number
  asset_code: string
  asset_name: string
  type: string
  expected: string
  actual: string
  cause: string
  action: string
}

export const auditApi = {
  getStats: () => apiFetch<AuditStats>('/api/v1/audit/stats', { token: token() }),
  listSessions: () => apiFetch<AuditSession[]>('/api/v1/audit/sessions', { token: token() }),
  createSession: (body: { location: string; audit_date: string }) =>
    apiFetch<AuditSession>('/api/v1/audit/sessions', { method: 'POST', token: token(), body }),
  getSession: (id: number) =>
    apiFetch<AuditSession>(`/api/v1/audit/sessions/${id}`, { token: token() }),
  updateSession: (id: number, body: { summary?: string; recommendation?: string; status?: string }) =>
    apiFetch<AuditSession>(`/api/v1/audit/sessions/${id}`, { method: 'PUT', token: token(), body }),
  saveRows: (id: number, rows: Omit<AuditRow, 'inspect_id' | 'session_id' | 'employee_id'>[]) =>
    apiFetch<AuditRow[]>(`/api/v1/audit/sessions/${id}/rows`, {
      method: 'POST', token: token(), body: { rows },
    }),
  addDiscrepancy: (id: number, body: Omit<Discrepancy, 'id' | 'session_id'>) =>
    apiFetch<Discrepancy>(`/api/v1/audit/sessions/${id}/discrepancies`, {
      method: 'POST', token: token(), body,
    }),
  deleteDiscrepancy: (discId: number) =>
    apiFetch<void>(`/api/v1/audit/discrepancies/${discId}`, { method: 'DELETE', token: token() }),
  submitReport: (id: number, note: string) =>
    apiFetch<AuditSession>(`/api/v1/audit/sessions/${id}/submit`, {
      method: 'POST', token: token(), body: { note },
    }),
  reviewReport: (id: number, action: 'approve' | 'reject', note: string) =>
    apiFetch<AuditSession>(`/api/v1/audit/sessions/${id}/review`, {
      method: 'POST', token: token(), body: { action, note },
    }),
}

const SESSION_KEY = 'audit_session_id'
export const activeSession = {
  get: (): number | null => {
    try {
      const v = sessionStorage.getItem(SESSION_KEY)
      return v ? Number(v) : null
    } catch {
      return null
    }
  },
  set: (id: number) => {
    try { sessionStorage.setItem(SESSION_KEY, String(id)) } catch {}
  },
  clear: () => {
    try { sessionStorage.removeItem(SESSION_KEY) } catch {}
  },
}
